### PR TITLE
Refactor k8s-handlers to make requests to kube-apiserver

### LINKED ISSUE #531: Use `client-go` to make requests to kube-apiserver instead of direct API calls
## Feature Request

Use `client-go` to make requests to kube-apiserver instead of direct API calls

**Is your feature request related to a problem? Please describe the use case.**

Currently we need to have a running proxy for kube-apiserver for KubeArmor in process mode to interact with k8s. 

**Describe the solution you'd like**

Current : https://github.com/kubearmor/KubeArmor/blob/60c105f04f1f3bf6fdc5d90264236e104c153d34/KubeArmor/core/k8sHandler.go#L238

Proposing: 
```
nodeWatcher, err := clientset.CoreV1().Nodes().Watch(metav1.ListOptions{})
```
Ref https://github.com/kubernetes/client-go#how-to-use-it

### PR DESCRIPTION
**Pull request by sbin64**
Closes #531

### CODE REVIEW

**Comment by sbin64** [comment]
@daemon1024 I made a few changes. Could you pull it and check it if it works? Thanks.

**Comment by daemon1024** [comment]
@afzalbin64 sorry for the delayed response! What's the current status of the PR? I wanted to know what should I be looking for and how can help out :) Thanks

**Comment by sbin64** [comment]
yup @daemon1024 alr8!

now you know, we are using `watch.Interface` type instead of `http.Response`, so `event.Object` doesn't access `Status.Addresses` as shown in the below against `Object` which is defined using `Object v1.Node json:"object"` previously.
https://github.com/kubearmor/KubeArmor/blob/2be0baf5cc114487025611d1b5fa6fc49b6ee797/KubeArmor/core/kubeUpdate.go#L75
https://github.com/kubearmor/KubeArmor/blob/2be0baf5cc114487025611d1b5fa6fc49b6ee797/KubeArmor/core/kubeUpdate.go#L86

**Comment by daemon1024** [comment]
Hey @afzal442 @afzalbin64,
I recently implemented a namespace watcher using client-go 
https://github.com/kubearmor/KubeArmor/blob/c21329ef3be1de3a72f180e612ce0c7ca28af240/KubeArmor/core/kubeUpdate.go#L1706-L1729

As part of https://github.com/kubearmor/KubeArmor/commit/c21329ef3be1de3a72f180e612ce0c7ca28af240. Hopefully this helps you out in implementing things here ✌️

**Comment by sbin64** [comment]
Got that @daemon1024. Thanks so much. I'll look into it this weekend. 😄

**Comment by afzal442** [comment]
@daemon1024 could u take a look here? I tested it locally through examples. It worked. Let me know what to do with `KubeArmor PodSecurityPolicies` Watcher. Any suggestion. Thanks

**Review by daemon1024 — CHANGES_REQUESTED** [review]
Couple of errors in the code causing the compilation of binary to fail, can you please fix it?
Can you try to apply a couple of policies and check if the enforcement is working or not to confirm if the code is fine or not. Ref https://github.com/kubearmor/KubeArmor/blob/main/contribution/testing_guide.md#1--test-kubearmor-in-manual

**Comment by daemon1024** [comment]
Thanks for working towards this. Please mark it ready for review once KubeArmor starts compiling/building. The changes seem to be in the right direction. 

Regarding how to deal with custom resources, you can find clients for them listed under https://github.com/kubearmor/KubeArmor/tree/main/pkg/KubeArmorPolicy/client. Ig the informer client will help you out but I am not really sure, so might need to look into it.

**Comment by sbin64** [comment]
My bad, sorry @daemon1024 , I think this last commit must have fixed the minor linting issue. But I'm not sure about KubeArmor `K8sSecurityPolicies` Watcher. Any suggestion. Thanks. 🛩️

**Comment by sbin64** [comment]
> Regarding how to deal with custom resources, you can find clients for them listed

Do you mean `clientset` will help in this regard ? https://github.com/kubearmor/KubeArmor/tree/main/pkg/KubeArmorPolicy/client/listers/security.kubearmor.com

File: KubeArmor/core/k8sHandler.go [hunk 1]
```diff
@@ -209,38 +210,49 @@ func (kh *K8sHandler) DoRequest(cmd string, data interface{}, path string) ([]by
...
-		req.Header.Add("Authorization", fmt.Sprintf("Bearer %s", kh.K8sToken))
+		if kh.InitInclusterAPIClient() {
+			watchNodes, err := kh.K8sClient.
+				CoreV1().
+				Nodes().
+				Watch(
+					context.Background(),
+					metav1.ListOptions{},
+				)
+			defer watchNodes.Stop()
```
Comment (daemon1024) [inline KubeArmor/core/k8sHandler.go]: Wouldn't this stop the watcher after returning it?
In the test scenario, KubeArmor is unable to get node information which means this is not working as expected I believe 🤔

**Comment by daemon1024** [comment]
@afzalbin64 This will https://github.com/kubearmor/KubeArmor/blob/86954a698ae2a70e5ec5265510201edfb9ca2a59/pkg/KubeArmorPolicy/client/informers/externalversions/security.kubearmor.com/v1/kubearmorpolicy.go#L25 I believe.`clientset` only allows Listing afaik.

File: KubeArmor/core/k8sHandler.go (hunk 1 above)
Comment (afzal442) [reply KubeArmor/core/k8sHandler.go]: Ah, I see, `defer` has called twice which overrides the watcher. Thanks @daemon1024 for helping me fix. 😄 ✌🏻

**Comment by nyrahul** [comment]
Any updates on this? It would be nice to see this handled as part of next release.

**Comment by afzal442** [comment]
Thanks @nyrahul! I would really love to see that. But, I have been able to cover the code refactor for `watch nodes` and `watch pods` so far.

**Comment by sbin64** [comment]
Closing as there is no progress. Feel free to refer it.

**Pull request closed without merging**